File: Tutorials/Codes/Modern_Control_Course/Chapter22/Lesson5/Chapter22_Lesson5.m
% Chapter22_Lesson5.m
%
% Practical constraints and limitations of state feedback:
% simulation of ideal, saturated, rate-limited, and noisy state feedback.
%
% Optional Simulink note:
% The final section creates a minimal Simulink model if Simulink is installed.

clear; clc; close all;

A = [0 1; 0 0];
B = [0; 1];
K = [6 5];          % desired continuous poles -2 and -3 for double integrator
umax = 1.0;
maxRate = 8.0;
dt = 0.001;
tf = 6.0;
t = 0:dt:tf;
x0 = [1.2; 0.0];

[xIdeal, uIdeal] = simulateCase(A, B, K, umax, maxRate, dt, t, x0, false, false);
[xCon, uCon]     = simulateCase(A, B, K, umax, maxRate, dt, t, x0, true, false);
[xNoise, uNoise] = simulateCase(A, B, K, umax, maxRate, dt, t, x0, true, true);

Acl = A - B*K;
Q = eye(2);
P = lyap(Acl', Q);
rhoMax = umax^2 / (K * inv(P) * K');

disp('Lyapunov matrix P:');
disp(P);
disp('Guaranteed no-saturation ellipsoid radius rho <= ');
disp(rhoMax);

figure;
plot(t, xIdeal(1,:), 'DisplayName', 'ideal x1'); hold on;
plot(t, xCon(1,:), 'DisplayName', 'constrained x1');
plot(t, xNoise(1,:), 'DisplayName', 'constrained + noise x1');
grid on; xlabel('time [s]'); ylabel('position state');
legend('Location', 'best');

figure;
plot(t, uIdeal, 'DisplayName', 'ideal u'); hold on;
plot(t, uCon, 'DisplayName', 'saturated/rate-limited u');
plot(t, uNoise, 'DisplayName', 'saturated/rate-limited/noisy u');
grid on; xlabel('time [s]'); ylabel('control input');
legend('Location', 'best');

% Optional: create a minimal Simulink model with State-Space, Gain, Saturation,
% and Scope blocks. This section is skipped if Simulink is unavailable.
if exist('simulink', 'file') == 4
    modelName = 'Chapter22_Lesson5_Simulink';
    new_system(modelName);
    open_system(modelName);

    add_block('simulink/Continuous/State-Space', [modelName '/Plant']);
    set_param([modelName '/Plant'], 'A', mat2str(A), 'B', mat2str(B), ...
        'C', mat2str(eye(2)), 'D', mat2str([0;0]));

    add_block('simulink/Math Operations/Gain', [modelName '/StateFeedbackGain']);
    set_param([modelName '/StateFeedbackGain'], 'Gain', mat2str(-K), ...
        'Multiplication', 'Matrix(K*u)');

    add_block('simulink/Discontinuities/Saturation', [modelName '/Saturation']);
    set_param([modelName '/Saturation'], 'UpperLimit', num2str(umax), ...
        'LowerLimit', num2str(-umax));

    add_block('simulink/Sinks/Scope', [modelName '/Scope']);
    save_system(modelName);
    disp(['Created optional Simulink model: ' modelName '.slx']);
end

function [x, u] = simulateCase(A, B, K, umax, maxRate, dt, t, x0, constrained, noisy)
    x = zeros(2, numel(t));
    u = zeros(1, numel(t));
    x(:,1) = x0;
    rng(4);

    for k = 1:(numel(t)-1)
        xMeas = x(:,k);
        if noisy
            xMeas = xMeas + [0.02; 0.03] .* randn(2,1);
        end

        uCmd = -K * xMeas;
        if constrained
            uSat = min(max(uCmd, -umax), umax);
            if k == 1
                uPrev = 0;
            else
                uPrev = u(k-1);
            end
            step = maxRate * dt;
            u(k) = min(max(uSat, uPrev - step), uPrev + step);
        else
            u(k) = uCmd;
        end

        f = @(xx) A*xx + B*u(k);
        q1 = f(x(:,k));
        q2 = f(x(:,k) + 0.5*dt*q1);
        q3 = f(x(:,k) + 0.5*dt*q2);
        q4 = f(x(:,k) + dt*q3);
        x(:,k+1) = x(:,k) + (dt/6)*(q1 + 2*q2 + 2*q3 + q4);
    end
    u(end) = u(end-1);
end

File: Tutorials/Codes/Modern_Control_Course/Chapter22/Lesson5/Chapter22_Lesson5.py
"""
Chapter22_Lesson5.py

Practical constraints and limitations of state feedback:
- ideal full-state feedback
- actuator amplitude saturation
- actuator rate limiting
- measurement noise amplification through K
- Lyapunov ellipsoid estimate for the no-saturation region

Dependencies:
    pip install numpy matplotlib
"""

import numpy as np
import matplotlib.pyplot as plt


def solve_continuous_lyapunov_by_kron(A: np.ndarray, Q: np.ndarray) -> np.ndarray:
    """
    Solve A.T P + P A = -Q using vectorization:
        vec(A.T P + P A) = (I kron A.T + A.T kron I) vec(P)
    This avoids requiring scipy.
    """
    n = A.shape[0]
    M = np.kron(np.eye(n), A.T) + np.kron(A.T, np.eye(n))
    p = np.linalg.solve(M, -Q.reshape(-1, order="F"))
    return p.reshape((n, n), order="F")


def saturate(value: float, lower: float, upper: float) -> float:
    return min(max(value, lower), upper)


def rate_limit(new_value: float, old_value: float, max_rate: float, dt: float) -> float:
    step = max_rate * dt
    return saturate(new_value, old_value - step, old_value + step)


def simulate(constrained: bool, noisy: bool):
    # Double integrator: position and velocity.
    A = np.array([[0.0, 1.0],
                  [0.0, 0.0]])
    B = np.array([[0.0],
                  [1.0]])

    # Pole placement for desired poles -2 and -3 gives K = [6, 5].
    K = np.array([[6.0, 5.0]])
    umax = 1.0
    max_rate = 8.0
    noise_std = np.array([0.02, 0.03])

    dt = 0.001
    tf = 6.0
    t = np.arange(0.0, tf + dt, dt)
    x = np.zeros((len(t), 2))
    u = np.zeros(len(t))
    x[0] = np.array([1.2, 0.0])

    rng = np.random.default_rng(4)
    for k in range(len(t) - 1):
        x_meas = x[k].copy()
        if noisy:
            x_meas += rng.normal(0.0, noise_std)

        u_cmd = float(-K @ x_meas)
        if constrained:
            u_sat = saturate(u_cmd, -umax, umax)
            u[k] = rate_limit(u_sat, u[k - 1] if k > 0 else 0.0, max_rate, dt)
        else:
            u[k] = u_cmd

        # Fourth-order Runge-Kutta for xdot = A x + B u, with u held constant.
        def f(xv):
            return A @ xv + (B[:, 0] * u[k])

        k1 = f(x[k])
        k2 = f(x[k] + 0.5 * dt * k1)
        k3 = f(x[k] + 0.5 * dt * k2)
        k4 = f(x[k] + dt * k3)
        x[k + 1] = x[k] + (dt / 6.0) * (k1 + 2 * k2 + 2 * k3 + k4)

    u[-1] = u[-2]
    return t, x, u, A, B, K


def lyapunov_no_saturation_radius(A: np.ndarray, B: np.ndarray, K: np.ndarray, umax: float):
    """
    For Acl = A - B K and Lyapunov ellipsoid E_rho = {x: x.T P x <= rho},
    saturation is certainly inactive if
        sqrt(rho * k P^{-1} k.T) <= umax.
    Thus rho <= umax^2 / (k P^{-1} k.T).
    """
    Acl = A - B @ K
    Q = np.eye(A.shape[0])
    P = solve_continuous_lyapunov_by_kron(Acl, Q)
    Pinv = np.linalg.inv(P)
    denom = float(K @ Pinv @ K.T)
    rho_max = umax * umax / denom
    return P, rho_max


if __name__ == "__main__":
    t1, x_ideal, u_ideal, A, B, K = simulate(constrained=False, noisy=False)
    t2, x_con, u_con, _, _, _ = simulate(constrained=True, noisy=False)
    t3, x_noise, u_noise, _, _, _ = simulate(constrained=True, noisy=True)

    P, rho = lyapunov_no_saturation_radius(A, B, K, umax=1.0)
    print("K =", K)
    print("Lyapunov P =")
    print(P)
    print("Guaranteed no-saturation ellipsoid radius rho <=", rho)

    plt.figure()
    plt.plot(t1, x_ideal[:, 0], label="ideal x1")
    plt.plot(t2, x_con[:, 0], label="constrained x1")
    plt.plot(t3, x_noise[:, 0], label="constrained+noise x1")
    plt.xlabel("time [s]")
    plt.ylabel("position state")
    plt.grid(True)
    plt.legend()
    plt.tight_layout()

    plt.figure()
    plt.plot(t1, u_ideal, label="ideal u")
    plt.plot(t2, u_con, label="saturated/rate-limited u")
    plt.plot(t3, u_noise, label="saturated/rate-limited/noisy u")
    plt.xlabel("time [s]")
    plt.ylabel("control input")
    plt.grid(True)
    plt.legend()
    plt.tight_layout()

    plt.show()
